src/library/export/trackexportdlg.cpp

```diff
@@ -1,63 +1,215 @@
-                               TrackExportWorker* worker)
+TrackExportDlg::TrackExportDlg(QWidget* parent,
+        UserSettingsPointer pConfig,
+        TrackPointerList& tracks,
+        Grantlee::Context* context)
         : QDialog(parent),
           Ui::DlgTrackExport(),
           m_pConfig(pConfig),
-          m_worker(worker) {
+          m_tracks(tracks),
+          m_worker(nullptr),
+          m_context(context) {
     setupUi(this);
+
+    QString lastExportDirectory = m_pConfig->getValue(
+            ConfigKey("[Library]", "LastTrackCopyDirectory"),
+            QStandardPaths::writableLocation(QStandardPaths::MusicLocation));
+    folderEdit->setText(lastExportDirectory);
+
+    m_worker = new TrackExportWorker(folderEdit->text(), m_tracks);
+
     connect(cancelButton,
             &QPushButton::clicked,
             this,
             &TrackExportDlg::cancelButtonClicked);
+    connect(startButton,
+            &QPushButton::clicked,
+            this,
+            &TrackExportDlg::slotStartExport);
     exportProgress->setMinimum(0);
     exportProgress->setMaximum(1);
     exportProgress->setValue(0);
-    statusLabel->setText("");
     setModal(true);
 
+    QHeaderView* headerView = statusTable->horizontalHeader();
+    headerView->resizeSection(0, 600);
+    headerView->resizeSection(1, 600);
+    headerView->resizeSection(2, 100);
+    headerView->setStretchLastSection(true);
+    statusTable->setHorizontalHeader(headerView);
+
+    connect(browseButton,
+            &QAbstractButton::clicked,
+            this,
+            &TrackExportDlg::slotBrowseFolder);
+    connect(folderEdit,
+            &QLineEdit::textChanged,
+            [=](const QString& x) {
+                Q_UNUSED(x);
+                updatePreview();
+            });
+    connect(comboPattern,
+            &QComboBox::currentTextChanged,
+            [this](const QString& x) {
+                Q_UNUSED(x);
+                updatePreview();
+            });
+
     connect(m_worker,
             &TrackExportWorker::progress,
             this,
             &TrackExportDlg::slotProgress);
+    connect(m_worker,
+            &TrackExportWorker::result,
+            this,
+            &TrackExportDlg::slotResult);
     connect(m_worker,
             &TrackExportWorker::askOverwriteMode,
             this,
             &TrackExportDlg::slotAskOverwriteMode);
     connect(m_worker,
             &TrackExportWorker::canceled,
             this,
-            &TrackExportDlg::cancelButtonClicked);
+            &TrackExportDlg::stopWorker);
+
+    if (m_tracks.isEmpty()) {
+        QMessageBox::warning(
+                nullptr,
+                tr("Export Error"),
+                tr("No files selected"),
+                QMessageBox::Ok,
+                QMessageBox::Ok);
+        hide();
+        accept();
+    }
+    updatePreview();
+}
+
+TrackExportDlg::~TrackExportDlg() {
+    if (m_context) {
+        delete m_context;
```

Comment:
The dialog should not delete the pointer if it has not created it. 
This if is redundant delete() has its own null check.

Reply:
I will add it to the documentation that the TrackExportDialog will take ownership of the provided context. Since the TrackExportDialog is running async and it is not clear for the caller when the dialog is closed and in fact, will never use the context again for anything else, transfer of the ownership is tho only sensable way here.